pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +241,289 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+					// with the new value of oldestSeenOffset. This is so the next cycle handled partially processed record properly.
+					commitRec = &kgo.Record{
+						Topic:       state.Commit.Topic,
+						Partition:   state.Commit.Partition,
+						Offset:      state.Commit.At - 1, // Previous commit's offset-1 means we expect the next cycle to start from the commit's offset.
+						Timestamp:   state.CommitRecTimestamp,
+						LeaderEpoch: state.Commit.LeaderEpoch,
+					}
+				} else if commitRec == nil {
+					// The commit offset should be the last record that was fully processed and not the first record that was not fully processed.
+					commitRec = lastRec
+				}
+			}
+			lastRec = rec
+		}
+	}
+
+	// Nothing was consumed from Kafka at all.
+	if firstRec == nil {
+		level.Info(partitionLogger).Log("msg", "no records were consumed")
+		return nil
+	}
+
+	// No records were processed for this cycle.
+	if lastRec == nil {
+		level.Info(partitionLogger).Log("msg", "nothing to commit due to first record fetched is from next cycle", "first_rec_offset", firstRec.Offset)
+		return nil
+	}
+
+	// All samples in all records were processed. We can commit the last record's offset.
+	if commitRec == nil {
+		commitRec = lastRec
+	}
+
+	var err error
+	numBlocks, err = builder.CompactAndUpload(ctx, b.uploadBlocks)
+	if err != nil {
+		return err
+	}
+
+	commitRecOffset := commitRec.Offset + 1 // offset+1 means everything up to (including) the offset was processed
+
+	// We should take the max of "seen till" offset. If the partition was lagging
+	// due to some record not being processed because of a future sample, we might be
+	// coming back to the same consume cycle again.
+	oldestSeenOffset := max(lastRec.Offset, state.OldestSeenOffset)
+	// Take the max of block max times because of the same reasons above.
+	lastBlockEnd := blockEnd
+	if lastBlockEnd.Before(state.LastBlockEnd) {
+		lastBlockEnd = state.LastBlockEnd
+	}
+
+	commit := kadm.Offset{
+		Topic:       commitRec.Topic,
+		Partition:   commitRec.Partition,
+		At:          commitRecOffset,
+		LeaderEpoch: commitRec.LeaderEpoch,
+		Metadata:    marshallCommitMeta(commitRec.Timestamp.UnixMilli(), oldestSeenOffset, lastBlockEnd.UnixMilli()),
+	}
+	if err := b.commitOffset(ctx, partitionLogger, b.cfg.Kafka.ConsumerGroup, commit); err != nil {
+		return err
+	}
+
+	// The new lag is the distance between fully processed offsets. If the partition's cycle was chopped into sections due to lagging,
+	// the cycle will continue consuming the next portion of the lag.
+	state.Lag -= commitRecOffset - firstRec.Offset
+
+	state.OldestSeenOffset = oldestSeenOffset
+	state.LastBlockEnd = lastBlockEnd
+	state.Commit = commit
+
+	return nil
+}
+
+func (b *BlockBuilder) seekPartition(partition int32, offset int64) {
+	offsets := map[string]map[int32]kgo.EpochOffset{
+		b.cfg.Kafka.Topic: {
+			partition: {
+				Epoch:  -1, // Epoch doesn't play any role outside the group consumption.
```

Comment:
doesn't epoch help detect when leaders get reelected and might help detect that the client is out of date?